File: Vector3.py
from math import sqrt, sin, cos, atan2

class Vector3:

    def __init__(self, x, y=None, w=None):
        self.x = x
        self.y = y if y is not None else x
        self.w = w if w is not None else 0

    def normalize(self):
        m = self.length()
        if m == 0:
            return self
        return self/m

    def dot(self, v):
        return self.x * v.x + self.y * v.y + self.w * v.w

    def dot2(self):
        return self.dot(self)

    def cross(self, v):
        return Vector3(self.y * v.w - self.w * v.y, self.w * v.x - self.x * v.w, self.x * v.y - self.y * v.x)
    
    def length(self):
        return sqrt(self.x**2 + self.y**2 + self.w**2)
    
    def angle(self):
        return atan2(self.y, self.x)
    
    def translate(self, v):
        return self + v
    
    def scale(self, s):
        return self * s
    
    def rotate(self, theta):
        return Vector3(self.x * cos(theta) - self.y * sin(theta), self.x * sin(theta) + self.y * cos(theta), self.w)

    # Overriding for comparisons to other Vector3
    def __eq__(self, v): 
        return self.x == v.x and self.y == v.y and self.w == v.w
    
    def __add__(self, b):
        return Vector3(self.x + b.x, self.y + b.y, self.w + b.w)
        
    def __sub__(self, b):
        return Vector3(self.x - b.x, self.y - b.y, self.w - b.w)

    def __neg__(self):
        return Vector3(-self.x, -self.y, -self.w)
    
    def __truediv__(self, b):
        return Vector3(self.x / b, self.y / b, self.w / b)
    
    def __mul__(self, b):
        return Vector3(self.x * b, self.y * b, self.w * b)
    
    def __str__(self):
        return "({},{},{})".format(self.x, self.y, self.w)

File: assets/utils/vector3.js
class Vector3 {
    constructor(x, y=null, w=null) {
        this.x = x;
        if (y != null) {
            this.y = y;
        }
        else {
            this.y = x;
        }
        if (w != null) {
            this.w = w;
        }
        else {
            this.w = 0;
        }
    }
    normalize() {
        var m = this.length();
        if (m == 0) {
            return this;
        }
        return this.div(m);
    };
    dot(v) {
        return this.x * v.x + this.y * v.y + this.w * v.w;
    };
    dot2() {
        return this.dot(this);
    };
    length() {
        return Math.sqrt(Math.pow(this.x, 2) + Math.pow(this.y, 2) + Math.pow(this.w, 2));
    };
    cross(v){
        return new Vector3(this.y * v.w - this.w * v.y, this.w * v.x - this.x * v.w, this.x * v.y - this.y * v.x)
    }
    angle() {
        return Math.atan2(this.y, this.x);
    };
    translate(v) {
        return this.add(v);
    };
    scale(n) {
        return this.mul(n);
    };
    rotate(theta) {
        return new Vector3(this.x * Math.cos(theta) - this.y * Math.sin(theta), this.x * Math.sin(theta) + this.y * Math.cos(theta), this.w);
    };
    equals(v) {
        return this.x == v.x && this.y == v.y && this.w == this.v.w;
    };
    add(v) {
        return new Vector3(this.x + v.x, this.y + v.y, this.w + v.w);
    };
    sub(v) {
        return new Vector3(this.x - v.x, this.y - v.y, this.w - v.w);
    };
    neg() {
        return new Vector3(-this.x, -this.y, -this.w);
    };
    div(n) {
        return new Vector3(this.x / n, this.y / n, this.w / n);
    };
    mul(n) {
        return new Vector3(this.x * n, this.y * n, this.w * n);
    };
    toString() {
        return "(".concat(this.x, ",").concat(this.y, ",").concat(this.x, ")");
    };
}